stdlib/public/core/StringObject.swift

```diff
@@ -23,6 +19,139 @@
 
 @_fixed_layout @usableFromInline
 internal struct _StringObject {
+  /*
+
+  On 64-bit platforms, the discriminator is the most significant 8 bits of the
+  bridge object.
+
+  ┌───────────────────╥─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┐
+  │ Form              ║  7  │  6  │  5  │  4  │  3  │  2  │  1  │  0  │
+  ╞═══════════════════╬═════╪═════╪═════╪═════╪═════╧═════╧═════╧═════╡
+  │ Immortal, Small   ║  1  │ASCII│  1  │  0  │      small count      │
+  ├───────────────────╫─────┼─────┼─────┼─────┼─────┬─────┬─────┬─────┤
+  │ Immortal, Large   ║  1  │  0  │  0  │  0  │  0  │ TBD │ TBD │ TBD │
+  ╞═══════════════════╬═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╡
+  │ Native            ║  0  │  0  │  0  │  0  │  0  │ TBD │ TBD │ TBD │
+  ├───────────────────╫─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┤
+  │ Shared            ║  0  │  0  │  0  │  0  │  1  │ TBD │ TBD │ TBD │
+  ├───────────────────╫─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┤
+  │ Exotic            ║  0  │  0  │  0  │  1  │  1  │ TBD │ TBD │ TBD │
+  ╞═══════════════════╬═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╡
+  │ ObjC, Contiguous  ║  0  │  1  │  0  │  0  │  1  │ TBD │ TBD │ TBD │
+  ├───────────────────╫─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┤
+  │ ObjC, Slow        ║  0  │  1  │  0  │  1  │  1  │ TBD │ TBD │ TBD │
+  └───────────────────╨─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┘
+
+  b3: isForeign (i.e., object payload isn't a biased pointer)
+  b4: isSlow
+  b5: isSmall
+  b6: isObjC/isASCII
+  b7: isImmortal
+
```

Comment:
Note: Terminology needs review here; it is likely that my labeling doesn't match yours. (I.e., these aren't intended as suggestions for renaming.)

The `Discriminator` struct isn't limited to 32-bit platforms: `SmallString` uses it everywhere. (It's difficult/costly to initialize a `StringObject` from raw bits on 32-bit platforms, so I changed `SmallString` to extract the discriminator directly rather than going through a conversion. This allowed me to get rid of a bunch of arch conditionals, but (depending on whether a shift instruction is actually emitted) it may be a *tiny* pessimization on 64-bit CPUs.)

I did not go the whole way though -- `StringObject` still uses full-word `Nibbles` masks on 64-bit platforms. But I hate all the duped constants, so I'd much prefer to switch `StringObject` to using the `Discriminator` abstraction too.